File: lab/policies.py
"""Policy checks for OAL lab train/eval pipeline."""
from __future__ import annotations

from typing import Any


ALLOWED_LICENSES = {"apache-2.0", "mit", "bsd-3-clause", "cc-by-4.0"}
BLOCKED_SOURCE_TOKENS = {"unknown", "leaked", "stolen", "unauthorized", "pirated"}


def validate_dataset_source(dataset: dict[str, Any]) -> tuple[bool, str]:
    license_name = str(dataset.get("license", "")).lower()
    source = str(dataset.get("source", "")).lower()

    if not license_name:
        return False, "dataset license missing"
    if license_name not in ALLOWED_LICENSES:
        return False, f"dataset license not allowed: {license_name}"
    if any(token in source for token in BLOCKED_SOURCE_TOKENS):
        return False, "dataset source violates policy"
    return True, "ok"


def validate_model_source(model: dict[str, Any]) -> tuple[bool, str]:
    source = str(model.get("source", "")).lower()
    allow_distill = bool(model.get("allow_distill", False))

    if any(token in source for token in BLOCKED_SOURCE_TOKENS):
        return False, "model source violates policy"
    if not allow_distill:
        return False, "model source disallows distillation"
    return True, "ok"


def validate_job_request(job: dict[str, Any]) -> tuple[bool, str]:
    dataset = job.get("dataset")
    model = job.get("base_model")

    if not isinstance(dataset, dict):
        return False, "dataset block missing"
    if not isinstance(model, dict):
        return False, "base_model block missing"

    ok, reason = validate_dataset_source(dataset)
    if not ok:
        return False, reason

    ok, reason = validate_model_source(model)
    if not ok:
        return False, reason

    return True, "ok"

File: lab/policies.ts
type ValidationResult = {
  ok: boolean;
  reason: string;
};

const ALLOWED_LICENSES = new Set(["apache-2.0", "mit", "bsd-3-clause", "cc-by-4.0"]);
const BLOCKED_SOURCE_TOKENS = ["unknown", "leaked", "stolen", "unauthorized", "pirated"];

function fail(reason: string): ValidationResult {
  return { ok: false, reason };
}

function pass(): ValidationResult {
  return { ok: true, reason: "ok" };
}

export function validateDatasetSource(dataset: Record<string, unknown>): ValidationResult {
  const license = String(dataset.license || "").toLowerCase();
  const source = String(dataset.source || "").toLowerCase();

  if (!license) {
    return fail("dataset license missing");
  }
  if (!ALLOWED_LICENSES.has(license)) {
    return fail(`dataset license not allowed: ${license}`);
  }
  if (BLOCKED_SOURCE_TOKENS.some((token) => source.includes(token))) {
    return fail("dataset source violates policy");
  }
  return pass();
}

export function validateModelSource(model: Record<string, unknown>): ValidationResult {
  const source = String(model.source || "").toLowerCase();
  const allowDistill = Boolean(model.allow_distill);

  if (BLOCKED_SOURCE_TOKENS.some((token) => source.includes(token))) {
    return fail("model source violates policy");
  }
  if (!allowDistill) {
    return fail("model source disallows distillation");
  }
  return pass();
}

export function validateJobRequest(job: Record<string, unknown>): ValidationResult {
  const dataset = job.dataset;
  const baseModel = job.base_model;

  if (!dataset || typeof dataset !== "object" || Array.isArray(dataset)) {
    return fail("dataset block missing");
  }
  if (!baseModel || typeof baseModel !== "object" || Array.isArray(baseModel)) {
    return fail("base_model block missing");
  }

  const datasetResult = validateDatasetSource(dataset as Record<string, unknown>);
  if (!datasetResult.ok) {
    return datasetResult;
  }

  const modelResult = validateModelSource(baseModel as Record<string, unknown>);
  if (!modelResult.ok) {
    return modelResult;
  }

  return pass();
}
